arch/risc-v/src/common/riscv_misaligned.c

```diff
@@ -0,0 +1,501 @@
+    uint16_t funct3 : 3;
+  } sw;
+
+  struct
+  {
+    uint16_t op     : 2;
+    uint16_t rs2    : 5;
+    uint16_t imm76  : 2; /* IMM bit [7:6] */
+    uint16_t imm52  : 4; /* IMM bit [5:2] */
+    uint16_t funct3 : 3;
+  } swsp;
+
+  struct
+  {
+    uint16_t op     : 2;
+    uint16_t rs2    : 3;
+    uint16_t imm76  : 2; /* IMM bit [7:6] */
+    uint16_t rs1    : 3;
+    uint16_t imm53  : 3; /* IMM bit [5:3] */
+    uint16_t funct3 : 3;
+  } sd;
+
+  struct
+  {
+    uint16_t op     : 2;
+    uint16_t rs2    : 5;
+    uint16_t imm86  : 3; /* IMM bit [8:6] */
+    uint16_t imm53  : 3; /* IMM bit [5:3] */
+    uint16_t funct3 : 3;
+  } sdsp;
+} riscv_insn_c_t;
+
+/* Normal instruction encoding */
+
+typedef union
+{
+  uint32_t insn;
+
+  /* Load */
+
+  struct
+  {
+    uint32_t op     : 7;
+    uint32_t rd     : 5;
+    uint32_t funct3 : 3;
+    uint32_t rs1    : 5;
+    uint32_t imm    : 12;
+  } l;
+
+  /* Store */
+
+  struct
+  {
+    uint32_t op     : 7;
+    uint32_t imm2   : 5;
+    uint32_t funct3 : 3;
+    uint32_t rs1    : 5;
+    uint32_t rs2    : 5;
+    uint32_t imm1   : 7;
+  } s;
+} rv_insn_t;
+
+typedef struct
+{
+  uint8_t *dest;
+  uint8_t *src;
+  int      len;
+  bool     sext;
+} riscv_insn_ctx_t;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: sext
+ *
+ * Description:
+ *   Sign extension a given bit width data
+ *
+ ****************************************************************************/
+
+static intptr_t sext(intptr_t v, uint32_t w)
+{
+  w = 8 * sizeof(intptr_t) - w;
+  return v << w >> w;
+}
+
+/****************************************************************************
+ * Name: decode_insn_compressed
+ *
+ * Description:
+ *   Try to decode a compressed instruction
+ *
+ ****************************************************************************/
+
+static bool decode_insn_compressed(uintptr_t *regs, riscv_insn_ctx_t *ctx)
+{
+#ifdef CONFIG_ARCH_RV_ISA_C
+  uint16_t in;
+  uint32_t imm;
+  riscv_insn_c_t insn;
+
+  /* Fetch instruction */
+
+  memcpy(&insn.insn, (void *)regs[REG_EPC], 2);
+
+  in = insn.insn & INSN_C_MASK;
+
+  switch (in)
+    {
+      case INSN_C_LW:
+        imm = insn.lw.imm2 << 2 | insn.lw.imm53 << 3 | insn.lw.imm6 << 6;
+        ctx->dest = (uint8_t *)&regs[REG_X8 + insn.lw.rd];
+        ctx->src  = (uint8_t *)regs[REG_X8 + insn.lw.rs1] + imm;
+        ctx->len  = 4;
+        ctx->sext = true;
+        break;
+
+      case INSN_C_LWSP:
+        imm = insn.lwsp.imm42 << 2 | insn.lwsp.imm5 << 5 |
+              insn.lwsp.imm76 << 6;
+        ctx->dest = (uint8_t *)&regs[insn.lwsp.rd];
+        ctx->src  = (uint8_t *)regs[REG_SP] + imm;
+        ctx->len  = 4;
+        ctx->sext = true;
+        break;
+
+#  ifdef CONFIG_ARCH_RV64
```

Comment:
why not support 64bit load/store on 32bit? But INSN_LD and INSN_SD support on 32bit.

Reply:
LD/SD is only available on RV64

Reply:
if so should we guard INSN_LD and INSN_SD with CONFIG_ARCH_RV64 too?

Reply:
OK, let's add the guard